### ISSUE TITLE
Sponge layers in 3D

### ISSUE THREAD

**Issue by ululi1970** [MUON-CFD/SOMAR#4]
Can't get them to work. I can apply Diri conditions on the inflow velocities to achieve the same
goal, but cannot drive the flow by nudging it with sponge layers.

**Comment by EdSantilli** [comment]
For the baroclinic velocity comparison against Masoud's work, I do use an explicit body forcing term and I also use the sponge to relax to the tidal velocity at the boundaries. I figure since I am applying body forcing, the sponge forcing should remain small. But my physical BCs are set to 1st order extrapolation so that the tidal forcing isn't doubled down. Does your 3D simulation blow up if you use 1st order extrap BCs as I do in BeamGenerationBCUtil::basicVelFuncBC?

**Comment by EdSantilli** [comment]
For the record, I just ran the beam generation problem with sponge forcing and:
1. extrap BCs + tidal body forcing
2. tidal BCs + no body forcing
3. extrap BCs + no body forcing
All three runs produce very similar (eyeball test) results after 4.5 tidal cycles.
On to 3D...

**Comment by EdSantilli** [comment]
**Bug found**: It appears tidalvel is forcing the fluid to move in horizontal ellipses, but fillVelSpongeLayerTarget is fixing v=0. This may be why the 3D code blows up, but the 2D code, which does not have v, does not. I will fix this and let you know what happens.

**Question**: What input file did you use to generate the movie in your talk? I'd like to run with exactly the same parameters as you did when your simulation blew up.

**Comment by ululi1970** [comment]
I am afraid that is not the case. It blows up with your BeamGenerationBCUtil.cpp. Also, you do not need bottom topography. The simplest case where it still blows up is 
with Cartesian Coordinates, Periodic in y, no stratification (b=0), no refinement, body forcing in x and sponge layers in x.  Basically, thake your 2D Beam generation case, set Nsq=0, and set geometry to 0 and run it in three D with ibc.XXXPeriodic=0 1 0. If you want, I can send you the input file...

**Comment by EdSantilli** [comment]
Yes, please push your input file so I can pull it. I cannot get the simulation to blow up.

**Comment by EdSantilli** [comment]
When you get a chance, also make sure your AMRNavierStokesFill.cpp is pushed.
BTW, have a safe ride home.

**Comment by ululi1970** [comment]
I pushed a input file that generates the blow up, and the BeamGenerationBCUtil.cpp I do not think that the modification to the latter make a difference re your version. I pushed my AMRNavierStokesFill.cpp a while ago.

**Comment by ululi1970** [comment]
Ed,
I think "bug" is not a simple i.j instead of j.i. I did some testing in a 2D channel with 
body force (U\* omega *cos (omega *t)) plus sponge layers. Everything proceeds fine until 
at t=pi/( omega) the flow reverses. At this point, w which was exactly =0 
starts developing a very small error (10^(-11)), after which everything blows up 
in a few time steps. I think this is why in your forcing term you calculate the
body force as (sin(omega *t_new)-sin(omega *t_old))/dt. 

The tiny mismatch between the velocity, the sponge and the extrapolation at 
the boundary causes the pressure solver to blow up (I imagine \int div u !=0). 

Maybe a more robust sponge layer should relax the vorticity to background (i.e. zero). 

Bottom line, I do not believe this is a coding bug.

**Comment by EdSantilli** [comment]
Unfortunately, I still cannot reproduce the bug. Even when I clone your exact code from the repo and change nothing. If I'm not mistaken, you blow up right away. Is that correct?

**Comment by ululi1970** [comment]
Hi Ed,
with a clean clone of ululi1970/DEM and this input file, run in 2D, I get a
blow starting after about 38 time steps.
Alberto

**Comment by EdSantilli** [comment]
Hi Alberto,

I am finally blowing up. I think the error is your value of ibc.spongeDtMultLo = 0.001. (Ditto on the high end.) The relaxation time is then 0.001*dt, which produces an enormous sponge force since spongeForce=(u_target-u_current)/relaxationTime. We are not limiting dt based on the sponge force term, and I think this is why the simulation is blowing up. Try setting ibc.spongeDtMultLo = O(1 or 10) 0.0 0.0. I typically use between 8 and 15.

For the record, I used tidalForce = U0 \* (sin(omega \* t_new)-sin(omega \* t_old))/dt as the forcing term so that a simple newVel = oldVel + dt \* tidalForce would apply the exact integral of the impulse over dt. Using forward Euler by setting tidalVel = U0 \* omega \* cos(omega \* t_old) seems to be stable too.
